Support transformers in remote inbox notifications

moon0326 opened the pull request:
Fixes #6936 

This PR adds `Transformer` concept to the Remote Inbox Notification.

I think the Remote Inbox Notification works very similarly to how an E.T.L (Extract, Transform, and Load) system works, but without a transformer. We extract a value from the database, compare them, and load a note. I figured adding transformers makes sense and it's flexible.

This is an example of transformer definition.

```
  {
    "slug": "wc-payments-qualitative-feedback",
     ....
    ],
    "rules": [
      {
        "type": "option",
        "transformers": [
            {
                "use": "dotNotation",
                "arguments": {
                    "path": "industry"
                }
            },
            {
                "use": "arrayColumn",
                "arguments": {
                    "key": "slug"
                }
            },
            {
                "use": "arraySearch",
                "arguments": {
                    "value": "cbd-other-hemp-derived-products"
                }
            }
        ],
        "option_name": "woocommerce_onboarding_profile",
        "operation": "!=",
        "value": "cbd-other-hemp-derived-products",
        "default": []
      }
    ]
  }
```

This example uses three transformers -- dotNotation, arrayColumn, and arraySearch.

We use dotNotation transformer to select `industry` array -> use arrayColumn to get the values -> search for `cbd-other-hemp-derived-products`.

**woocommerce_onboarding_profile** value

```
Array
(
    [setup_client] => 
    [industry] => Array
        (
            [0] => Array
                (
                    [slug] => food-drink
                )

            [1] => Array
                (
                    [slug] => fashion-apparel-accessories
                )

            [2] => Array
                (
                    [slug] => health-beauty
                )

            [3] => Array
                (
                    [slug] => other
                    [detail] => test
                )

        )

    [product_types] => Array
        (
            [0] => downloads
        )

    [product_count] => 0
    [selling_venues] => no
    [business_extensions] => Array
        (
        )

    [theme] => twentytwentyone
    [completed] => 1
)
```

The [$option_value](https://github.com/woocommerce/woocommerce-admin/pull/6948/files#diff-d1413e004b7819467e9e6b12e3d4b0b19f27d4f719187bf828b8797a21b55fafR45) would be an array of selected industries since `arraySearch` transformer wasn't able to find `cbd-other-hemp-derived-products`

```
(
    [0] => food-drink
    [1] => fashion-apparel-accessories
    [2] => health-beauty
    [3] => other
)
```


### Detailed test instructions:

1. Prepare woocommerce.com local test environment.
2. Once your woocommerce.com local test environment is ready, open `plugins/wccom-plugins/inbox-notifications/includes/notifications.json.php` and add the following test note definition.

```
[
  {
    "slug": "test-note",
    "type": "info",
    "status": "unactioned",
    "is_snoozable": 0,
    "source": "woocommerce.com",
    "locales": [
      {
        "locale": "en_US",
        "title": "test note",
        "content": "test note"
      }
    ],
    "actions": [
      {
        "name": "test note",
        "locales": [
          {
            "locale": "en_US",
            "label": "Share feedback"
          }
        ],
        "url": "https://automattic.survey.fm/wc-pay-new",
        "url_is_admin_query": false,
        "is_primary": true,
        "status": "actioned"
      }
    ],
    "rules": [
      {
        "type": "option",
        "transformers": [
            {
                "use": "dot_notation",
                "arguments": {
                    "path": "industry"
                }
            },
            {
                "use": "array_column",
                "arguments": {
                    "key": "slug"
                }
            },
            {
                "use": "array_search",
                "arguments": {
                    "value": "fashion-apparel-accessories"
                }
            }
        ],
        "option_name": "woocommerce_onboarding_profile",
        "operation": "=",
        "value": "fashion-apparel-accessories",
        "default": []
      }
    ]
  }
]
```

3. Open `src/RemoteInboxNotifications/DataSourcePoller.php` in WC Admin and replace `DATA_SOURCES` to use your `woocommerce.test` local environment. 
4. Start OBW and choose 'Fashion, apparel, and accessories from the Industry selection.
5. Install [WP Crontrol 
](https://wordpress.org/plugins/wp-crontrol/)
6. Navigate to Tools -> Cron Events and run `wp_admin_daily`
7. Navigate to WooCommerce -> Home and confirm that the `test note` has been added.

moon0326 commented:
Hey @psealock @becdetat, this is my draft version of PR for the issue https://github.com/woocommerce/woocommerce-admin/issues/6936

I've added `Transformer` concept, but I wanted to hear some thoughts before I finalize them. Could you take a look at the description when you get a chance, please?

becdetat commented:
How does this work?
> The $option_value would be an array of selected industries since arraySearch transformer wasn't able to find cbd-other-hemp-derived-products

It looks like the `ArraySearch::transform` function would return `null` if `$value` isn't found?

```
public function transform( $value ) {
	$key = array_search( $this->value, $value, true );
	if ( false !== $key ) {
		return $value[ $key ];
	}

	return null;
}
```

becdetat commented:
Is this the equivalent of your example? It seems to me that you could drop the `arraySearch` and just use `contains`.

```
  {
    "slug": "wc-payments-qualitative-feedback",
     ....
    ],
    "rules": [
      {
        "type": "option",
        "transformers": [
            {
                "use": "dotNotation",
                "arguments": {
                    "path": "industry"
                }
            },
            {
                "use": "arrayColumn",
                "arguments": {
                    "key": "slug"
                }
            }
        ],
        "option_name": "woocommerce_onboarding_profile",
        "operation": "contains",
        "value": "cbd-other-hemp-derived-products",
        "default": []
      }
    ]
  }
```

moon0326 commented:
> How does this work?
> 
> > The $option_value would be an array of selected industries since arraySearch transformer wasn't able to find cbd-other-hemp-derived-products
> 
> It looks like the `ArraySearch::transform` function would return `null` if `$value` isn't found?
> 
> ```
> public function transform( $value ) {
> 	$key = array_search( $this->value, $value, true );
> 	if ( false !== $key ) {
> 		return $value[ $key ];
> 	}
> 
> 	return null;
> }
> ```

So the current logic is to return `$target_value` if a transformer returns null.  See [here](https://github.com/woocommerce/woocommerce-admin/pull/6948/files#diff-87f9cce1aec0a6026b524f95d3f3389be38662a479e16f0771cd3e3471987c3aR58).

moon0326 commented:
> > How does this work?
> > > The $option_value would be an array of selected industries since arraySearch transformer wasn't able to find cbd-other-hemp-derived-products
> > 
> > 
> > It looks like the `ArraySearch::transform` function would return `null` if `$value` isn't found?
> > ```
> > public function transform( $value ) {
> > 	$key = array_search( $this->value, $value, true );
> > 	if ( false !== $key ) {
> > 		return $value[ $key ];
> > 	}
> > 
> > 	return null;
> > }
> > ```
> 
> So the current logic is to return `$target_value` if a transformer returns null. See [here](https://github.com/woocommerce/woocommerce-admin/pull/6948/files#diff-87f9cce1aec0a6026b524f95d3f3389be38662a479e16f0771cd3e3471987c3aR58).

Good catch 👍 Yup, I think we can also use contains in this case.

becdetat commented:
I'm not sure about using camelCase for the transformer `use` value. Rule types are snack_case, IMO it would be more consistent to use that for transformers as well. I like the factory that you built but it could probably convert snake_case to PascalCase instead.

moon0326 commented:
> > > How does this work?
> > > > The $option_value would be an array of selected industries since arraySearch transformer wasn't able to find cbd-other-hemp-derived-products
> > > 
> > > 
> > > It looks like the `ArraySearch::transform` function would return `null` if `$value` isn't found?
> > > ```
> > > public function transform( $value ) {
> > > 	$key = array_search( $this->value, $value, true );
> > > 	if ( false !== $key ) {
> > > 		return $value[ $key ];
> > > 	}
> > > 
> > > 	return null;
> > > }
> > > ```
> > 
> > 
> > So the current logic is to return `$target_value` if a transformer returns null. See [here](https://github.com/woocommerce/woocommerce-admin/pull/6948/files#diff-87f9cce1aec0a6026b524f95d3f3389be38662a479e16f0771cd3e3471987c3aR58).
> 
> Good catch 👍 Yup, I think we can also use contains in this case.



> > > How does this work?
> > > > The $option_value would be an array of selected industries since arraySearch transformer wasn't able to find cbd-other-hemp-derived-products
> > > 
> > > 
> > > It looks like the `ArraySearch::transform` function would return `null` if `$value` isn't found?
> > > ```
> > > public function transform( $value ) {
> > > 	$key = array_search( $this->value, $value, true );
> > > 	if ( false !== $key ) {
> > > 		return $value[ $key ];
> > > 	}
> > > 
> > > 	return null;
> > > }
> > > ```
> > 
> > 
> > So the current logic is to return `$target_value` if a transformer returns null. See [here](https://github.com/woocommerce/woocommerce-admin/pull/6948/files#diff-87f9cce1aec0a6026b524f95d3f3389be38662a479e16f0771cd3e3471987c3aR58).
> 
> Good catch 👍 Yup, I think we can also use contains in this case.



> I'm not sure about using camelCase for the transformer `use` value. Rule types are snack_case, IMO it would be more consistent to use that for transformers as well. I like the factory that you built but it could probably convert snake_case to PascalCase instead.

That makes sense. I missed it. I will update `use` value to use snake_case.

becdetat commented:
Unit tests around the transform functions would be great :-D

moon0326 commented:
> Unit tests around the transform functions would be great :-D

Yup 👍 that's the plan. I'm gonna update README.md with transformers, add a few more array-related transformers, and write some tests to wrap it up.

moon0326 commented:
> Unit tests around the transform functions would be great :-D

I've added tests and updated README.md. Could you take a look at them when you get a chance, please? 🙇‍♂️

becdetat commented on src/RemoteInboxNotifications/TransformerService.php:
```diff
@@ -0,0 +1,69 @@
+	 */
+	public static function apply( $target_value, array $transformer_configs ) {
+		foreach ( $transformer_configs as $transformer_config ) {
+			if ( ! isset( $transformer_config->use ) ) {
```
This could be done in `OptionRuleProcessor`'s `validate` function - any invalid transforms should invalidate the rule

moon0326 replied:
I think it's better if we catch exception from the `Option` class instead of validating the config in the `Option`. What do you think?

we could simply try and catch and invalidate the rule in case of any exception thrown from the `TransformerService`

becdetat replied:
The idea of spec validation was to catch invalid specs before they got imported (during polling), rather than during execution. Plus, exceptions shouldn't be used for flow control.

moon0326 replied:
Looking at the flow of how the whole Inbox Remote Notification system works again, you're correct. I will make an update soon 👍

becdetat commented on src/RemoteInboxNotifications/Transformers/ArrayColumn.php:
```diff
@@ -0,0 +1,32 @@
+	 * @return mixed
+	 */
+	public function transform( $value, stdClass $arguments = null ) {
+		if ( ! isset( $arguments->key ) ) {
```
Again, this could be in the rule processor's `validate` function

becdetat replied:
alternatively there could be a `validate` function on the transformer classes, and `OptionRuleProcessor` could just call into the appropriate implementation.

moon0326 replied:
That's a good suggestion. I think I will add `validate` in each Transformer class as they should protect themselves.

moon0326 replied:
That's a good point, but I think an object should protect itself again any values coming it. It's also going to be useful in case we ever want to use Transformers in other types other than Option. We wouldn't have to repeat code (calling validate)

becdetat replied:
I agree that the transformer is the appropriate place for validation to happen, I just think it could be in a seperate method and called as part of rule validation.

> We wouldn't have to repeat code (calling validate)

The only code that would be repeated would be like (inside of `OptionRuleTransformer`:
```
foreach ( $rule->transformers as $transform_args ) {
    $transformer = TrasformerService::create_transformer( $transform_args->use );
    if ( ! $transformer->validate( $transform_args->arguments ) ) {
        return false;
    }
}
```

Then if and when this needs to be duplicated, we look at deduplicating it then.

moon0326 replied:
Looking at the actual code, I think we can do that. That wouldn't be too hard to copy and paste in case we ever need to. I will make an update soon 👍

becdetat commented on src/RemoteInboxNotifications/Transformers/README.md:
```diff
@@ -0,0 +1,169 @@
+
+Arguments:
+
+|name|descriont|
```
is this a typo?

moon0326 replied:
Yup 😄 I will fix it.

becdetat commented on tests/remote-inbox-notifications/transformer-service.php:
```diff
@@ -0,0 +1,116 @@
+	 * When the second transformer returns null
+	 * Then the value transformed by the first transformer should be returned.
+	 */
+	public function test_it_returns_previously_transformed_value_when_transformer_returns_null() {
```
loving these tests

moon0326 replied:
Thank you! I like this approach too. It helps me to organize my tests before writing them :)

becdetat commented:
This looks great. As I commented I just have an issue with exceptions being used for validation. Also to make testing easier you can create a public gist with a feed JSON in it and add the raw URL to `DataSourcePoller`.

moon0326 commented:
> This looks great. As I commented I just have an issue with exceptions being used for validation. Also to make testing easier you can create a public gist with a feed JSON in it and add the raw URL to `DataSourcePoller`.

Thank you! I will make updates on the validation parts. 

Using a public gist is brilliant! never thought of that 👍 👍

moon0326 commented:
@becdetat Added validate() in https://github.com/woocommerce/woocommerce-admin/pull/6948/commits/9250fde5efc172979fe90ed8f217c930ea350e04

becdetat reviewed (approved):
Thanks @moon0326 , this looks great and works well
